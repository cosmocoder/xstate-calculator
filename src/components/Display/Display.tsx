import type { CalcData } from '../../types';

export interface DisplayProps {
  currentCalc: CalcData;
}

const calculateFontSize = (numLength: number) => {
  const excess = numLength - 6;
  const calculatedSize = 50 - excess * 3.5;
  const fontSize = calculatedSize > 18 ? calculatedSize : 18;
  return `${fontSize}px`;
};

export const Display = ({ currentCalc }: DisplayProps) => {
  const { previousNum, currentNum, result } = currentCalc;

  const numToDisplay = result || currentNum || previousNum || '0';

  const numLength = numToDisplay.toString().length;

  const style = numLength > 7 ? { fontSize: calculateFontSize(numLength) } : undefined;

  return (
    <div style={style} className="display">
      {numToDisplay}
    </div>
  );
};
