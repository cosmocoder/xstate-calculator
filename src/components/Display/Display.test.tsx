import { render, screen } from '@testing-library/react';
import { Display, DisplayProps } from './Display';

const defaultProps: DisplayProps = {
  currentCalc: {
    currentNum: '',
    previousNum: '',
    operation: undefined,
    result: '',
  },
};

describe('Display', () => {
  it('should render correctly in default state', async () => {
    const { container } = render(<Display {...defaultProps} />);

    expect(container).toMatchSnapshot();
    expect(await screen.findByText('0')).toBeInTheDocument();
  });

  it('should render correctly when only currentNum is available', async () => {
    render(<Display currentCalc={{ ...defaultProps.currentCalc, currentNum: '2' }} />);

    expect(await screen.findByText('2')).toBeInTheDocument();
  });

  it('should render correctly when both currentNum and previousNum are available', async () => {
    render(<Display currentCalc={{ ...defaultProps.currentCalc, currentNum: '3', previousNum: '2' }} />);

    expect(await screen.findByText('3')).toBeInTheDocument();
  });

  it('should render correctly when both result is available', async () => {
    render(<Display currentCalc={{ ...defaultProps.currentCalc, currentNum: '3', previousNum: '2', result: '5' }} />);

    expect(await screen.findByText('5')).toBeInTheDocument();
  });
});
