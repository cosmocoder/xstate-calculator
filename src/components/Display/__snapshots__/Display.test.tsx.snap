// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Display > should render correctly in default state 1`] = `
<div>
  <div
    class="display"
  >
    0
  </div>
</div>
`;
